test/unified_format.py

```diff
@@ -0,0 +1,999 @@
+                                     replace(' ', '_').replace('.', '_'),)
+            test_method = create_test(copy.deepcopy(test_spec))
+            test_method.__name__ = test_name
+
+            if description in cls.EXPECTED_FAILURES:
+                test_method = unittest.expectedFailure(test_method)
+
+            setattr(cls, test_name, test_method)
+
+
+_ALL_MIXIN_CLASSES = [
+    UnifiedSpecTestMixinV1,
+    # add mixin classes for new schema major versions here
+]
+
+
+_SCHEMA_VERSION_MAJOR_TO_MIXIN_CLASS = {
+    KLASS.SCHEMA_VERSION[0]: KLASS for KLASS in _ALL_MIXIN_CLASSES}
+
+
+def generate_test_classes(test_path, module=__name__, class_name_prefix='',
+                          expected_failures=[]):
+    """Method for generating test classes. Returns a dictionary where keys are
+    the names of test classes and values are the test class objects."""
+    test_klasses = {}
+
+    def test_base_class_factory(test_spec):
+        """Utility that creates the base class to use for test generation.
+        This is needed to ensure that cls.TEST_SPEC is appropriately set when
+        the metaclass __init__ is invoked."""
+        class SpecTestBase(with_metaclass(UnifiedSpecTestMeta)):
+            TEST_SPEC = test_spec
+            EXPECTED_FAILURES = expected_failures
+        return SpecTestBase
+
+    for dirpath, _, filenames in os.walk(test_path):
+        dirname = os.path.split(dirpath)[-1]
+
+        for filename in filenames:
+            fpath = os.path.join(dirpath, filename)
+            with open(fpath) as scenario_stream:
+                # Use tz_aware=False to match how CodecOptions decodes
+                # dates.
+                opts = json_util.JSONOptions(tz_aware=False)
+                scenario_def = json_util.loads(
+                    scenario_stream.read(), json_options=opts)
+
+            test_type = os.path.splitext(filename)[0]
+            snake_class_name = 'Test%s_%s_%s' % (
+                class_name_prefix, dirname.replace('-', '_'),
+                test_type.replace('-', '_').replace('.', '_'))
+            class_name = snake_to_camel(snake_class_name)
+
+            try:
+                schema_version = Version.from_string(
+                    scenario_def['schemaVersion'])
+                mixin_class = _SCHEMA_VERSION_MAJOR_TO_MIXIN_CLASS.get(
+                    schema_version[0])
+                if mixin_class is None:
+                    print('Ignoring test file %s with '
+                          'unsupported schemaVersion %s' %
+                          (fpath, schema_version))
+                    continue
+                test_klasses[class_name] = type(
+                    class_name,
+                    (mixin_class, test_base_class_factory(scenario_def),),
+                    {'__module__': module})
+            except (AttributeError, KeyError, TypeError):
+                print("Ignoring invalid test file '%s'\n"
```

Comment:
Shouldn't we raise some kind of error here so that we actually notice when this happens?

Reply:
Same answer as above.